mersenne-31/src/poseidon2.rs

```diff
@@ -0,0 +1,201 @@
+use p3_field::{AbstractField, Field, PrimeField32};
+use p3_poseidon2::DiffusionPermutation;
+use p3_symmetric::Permutation;
+
+use crate::{to_mersenne31_array, Mersenne31};
+
+// Two optimised diffusion matrices for Mersenne31/16:
+
+// Mersenne31:
+// Small entries: [-2, 1, 2, 3, 4, 5, 6, 7, 8, 9, 12, 13, 14, 15, 16, 17]
+// Power of 2 entries: [-2,  1,   2,   4,   8,  16,  32,  64, 128, 256, 1024, 4096, 8192, 16384, 32768, 65536]
+//                   = [?, 2^0, 2^1, 2^2, 2^3, 2^4, 2^5, 2^6, 2^7, 2^8, 2^10, 2^12, 2^13,  2^14,  2^15, 2^16]
+
+const MATRIX_DIAG_16_MERSENNE31_U32: [u32; 16] = [
+    Mersenne31::ORDER_U32 - 2,
+    1,
+    2,
+    4,
+    8,
+    16,
+    32,
+    64,
+    128,
+    256,
+    1024,
+    4096,
+    8192,
+    16384,
+    32768,
+    65536,
+];
+
+pub const MATRIX_DIAG_16_MERSENNE31: [Mersenne31; 16] =
+    to_mersenne31_array(MATRIX_DIAG_16_MERSENNE31_U32);
+
+// We should do some sort of delayed reduction strategy using the shifts.
+const MATRIX_DIAG_16_MONTY_SHIFTS: [i32; 16] =
+    [-64, 0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 15, 16];
+// Note that the first entry of this constant should never be accessed.
+
+#[derive(Debug, Clone, Default)]
+pub struct DiffusionMatrixMersenne31;
+
+impl Permutation<[Mersenne31; 16]> for DiffusionMatrixMersenne31 {
+    fn permute_mut(&self, state: &mut [Mersenne31; 16]) {
+        let sum: Mersenne31 = state.iter().cloned().sum();
+        state[0] = sum - state[0].double();
+        for i in 1..16 {
+            state[i] = state[i].mul_2exp_u64(MATRIX_DIAG_16_MONTY_SHIFTS[i] as u64);
+            state[i] += sum;
+        }
+    }
+}
```

Comment:
As discussed, this is a bit cleaner:
```suggestion
const MATRIX_DIAG_16_MONTY_SHIFTS: [i32; 15] =
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 15, 16];
// Note that the first entry of this constant should never be accessed.

#[derive(Debug, Clone, Default)]
pub struct DiffusionMatrixMersenne31;

impl Permutation<[Mersenne31; 16]> for DiffusionMatrixMersenne31 {
    fn permute_mut(&self, state: &mut [Mersenne31; 16]) {
        let sum: Mersenne31 = state.iter().cloned().sum();
        state[0] = sum - state[0].double();
        for i in 1..16 {
            state[i] = state[i].mul_2exp_u64(MATRIX_DIAG_16_MONTY_SHIFTS[i-1] as u64);
            state[i] += sum;
        }
    }
}
```